### PR TITLE
python27: switch to ActiveState's fork for Python 2

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Description of changes

ActiveState is a company that is maintaining a fork of Python 2 to fixes its security issues. Their support is paid, however the code is open-source. See the details here: https://www.activestate.com/products/python/python-2-end-of-life-security-updates/

This enable us to drop a bunch of CVE's patches for Python 2.7 and also it should be easier to maintain, since we can just bump the version once ActiveState tags a new version.

See the discussion of https://github.com/NixOS/nixpkgs/pull/201859 for the reasoning.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 1]
```diff
@@ -119,14 +121,6 @@ let
       # Backport from CPython 3.8 of a good list of tests to run for PGO.
       ./profile-task.patch
 
-      # https://www.activestate.com/products/python/python-2-end-of-life-security-updates/
-      ./CVE-2019-20907.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:123]: https://github.com/ActiveState/cpython/commit/e8efbb448a26b37d511a697c272b71fbad41fe77

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 2]
```diff
@@ -119,14 +121,6 @@ let
...
-      ./CVE-2020-8492.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:124]: https://github.com/ActiveState/cpython/commit/33e2642505c188ec168464e8ddfd49d3ed8181f8

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 3]
```diff
@@ -119,14 +121,6 @@ let
...
-      ./CVE-2020-26116.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:125]: https://github.com/ActiveState/cpython/commit/15b0c6352e2d1620ee9f6973be0e099bf64f59af

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 4]
```diff
@@ -119,14 +121,6 @@ let
...
-      ./CVE-2020-27619.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:126]: https://github.com/ActiveState/cpython/commit/ec1b4657267074cbe5b7c4a549b0e689bb51281a

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 5]
```diff
@@ -119,14 +121,6 @@ let
...
-      ./CVE-2021-3177.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:127]: https://github.com/ActiveState/cpython/commit/21f322e1953a66004dc6df413552936103baed29

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 6]
```diff
@@ -119,14 +121,6 @@ let
...
-      ./CVE-2021-23336.patch
```
Comment (thiagokokada) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:128]: https://github.com/ActiveState/cpython/commit/dbdd1349d1972e9a3c788c2af060574c1e36323c

**Comment by thiagokokada** [comment]
BTW, this will fix CVE-2022-0391 (see https://github.com/ActiveState/cpython/commit/8c70fb47fee85eb8ea926aa3a2ae57a11c04b006), that was not fixed in #203362, so we should merge it ASAP if this is the way we want to go.

**Comment by thiagokokada** [comment]
In `master` branch of ActiveState there one other CVE that will be included in the 2.7.18.6 release: https://github.com/ActiveState/cpython/commit/eeb7fe50450f08a782921f3229abed2f23e7b2d7

However, this CVE-2021-3733 is a ReDOS bug so it shouldn't be that problematic. And it should be fixed soon anyway (once ActiveState tags a new release).

**Comment by thiagokokada** [comment]
> In `master` branch of ActiveState there is one other CVE that will be included in the 2.7.18.6 release: [ActiveState/cpython@eeb7fe5](https://github.com/ActiveState/cpython/commit/eeb7fe50450f08a782921f3229abed2f23e7b2d7)
> 
> However, this [CVE-2021-3733](https://github.com/advisories/GHSA-wj55-vqcq-gxcp) is a DoS bug so it shouldn't be that problematic. And it should be fixed soon anyway (once ActiveState tags a new release).

Decided to backport this patch. Now all known issues from ActiveState should be fixed.

**Review by FRidh — APPROVED** [review]
I think this is a good idea.

I have not tested the build but by reading what it is about and skimming the commits they added I see no reason why there would be any issues.

**Review by mweinelt — APPROVED** [review]

**Comment by fabianhjr** [comment]
Hey, I remember some time ago reading that the Python foundation made a public position that they would send c&d to forks of python 2 that created new tags (like 2.7.18.5) nad didn't differentiate the name enough from python.

**Comment by thiagokokada** [comment]
> Hey, I remember some time ago reading that the Python foundation made a public position that they would send c&d to forks of python 2 that created new tags (like 2.7.18.5) nad didn't differentiate the name enough from python.

I wouldn't worry it too much right now, unless Python itself started to enforce it.

In the worst case, we can patch ActiveState to show `2.7.18` instead. If I remember correctly, it is a one file patch.

**Comment by fabianhjr** [comment]
A thread about trademark anda "Python 2.8" fork.

https://github.com/naftaliharris/tauthon/issues/47

**Comment by thiagokokada** [comment]
> A thread about trademark anda "Python 2.8" fork.
> 
> [naftaliharris/tauthon#47](https://github.com/naftaliharris/tauthon/issues/47)

`tauthon` is very much a different beast though. It add new features and even has a few breaking changes compared to Python 2.7, so it is pretty much not Python.

ActivateState's fork is just a few security fixes, no new features. It is not that much different from us shipping a few extra patches with the exception of the add prefix to the version. If Python devs complain about it, I can happily patch the version patches out. Otherwise let's just keep it as-is.

**Comment by fabianhjr** [comment]
Sounds good

**Pull request merged**